src/imitation/policies/replay_buffer_wrapper.py

```diff
@@ -0,0 +1,90 @@
+"""Wrapper for reward labeling for transitions sampled from a replay buffer."""
+
+
+from typing import Mapping, Type
+
+import numpy as np
+import torch as th
+from gym import spaces
+from stable_baselines3.common.buffers import ReplayBuffer
+from stable_baselines3.common.type_aliases import ReplayBufferSamples
+
+from imitation.rewards.reward_function import RewardFn
+from imitation.util import util
+
+
+def _samples_to_reward_fn_input(
+    samples: ReplayBufferSamples,
+) -> Mapping[str, np.ndarray]:
+    """Convert a sample from a replay buffer to a numpy array."""
+    return dict(
+        state=samples.observations.cpu().numpy(),
+        action=samples.actions.cpu().numpy(),
+        next_state=samples.next_observations.cpu().numpy(),
+        done=samples.dones.cpu().numpy(),
+    )
+
+
+class ReplayBufferRewardWrapper(ReplayBuffer):
+    """Relabel the rewards in transitions sampled from a ReplayBuffer."""
+
+    def __init__(
```

Comment:
`__init__` is failing to set several attributes that would usually be set by `BaseBuffer`: namely `buffer_size`, `observation_space`, `action_space`, ...

I get that we can't call `super().__init__` directly because `ReplayBuffer.__init__` sets some things we don't want (e.g. `self.observations`), but perhaps it's worth adding a call to `BaseBuffer.__init__` directly?

Bit hacky but seems better than not calling `__init__` at all.

Thoughts @tomtseng?

Reply:
`BaseBuffer.__init__` seems better than `ReplayBuffer.__init__`. Foreseeable problems might be `self.pos` and `self.full` will not be incremented by `self.add()`. One way of workaround would be adding the following into `self.add()`:

```python
self.pos += 1
if self.pos == self.buffer_size:
    self.full = True
    self.pos = 0
```

But not sure whether this is desirable.

Reply:
FWIW, `BaseBuffer` has an abstractmethod `_get_samples()` that's not used anywhere in the wrapper, so in practice we could do `def _get_samples(): NotImplementedError`

Reply:
Not sure if this is appropriate, but another (verbose) workaround would be to use `@property` decorators like
```py
  @property
  def pos(self):
    return self.replay_buffer.pos
```
This is what `gym` does with their environment wrappers — the `Wrapper` class  is a subclass of `Env`, and instead of calling `super().__init__()`, `Wrapper` has a bunch of property decorators like
```py
    @property
    def action_space(self) -> spaces.Space[ActType]:
        """Returns the action space of the environment."""
        if self._action_space is None:
            return self.env.action_space
        return self._action_space
```
Then all environment wrappers inherit can subclass `Wrapper`, and calling `super().__init__(env)` behaves nicely

Reply:
Gotcha, this helps!